agent/engine/docker_task_engine.go

```diff
@@ -1238,12 +1240,139 @@ func (engine *DockerTaskEngine) GetDaemonManagers() map[string]dm.DaemonManager
 func (engine *DockerTaskEngine) pullContainerManifest(
 	task *apitask.Task, container *apicontainer.Container,
 ) dockerapi.DockerContainerMetadata {
-	// Currently a no-op
-	logger.Debug("Manifest pull is currently a no-op", logger.Fields{
-		field.TaskID:    task.GetID(),
-		field.Container: container.Name,
-		field.Image:     container.Image,
+	switch container.Type {
+	case apicontainer.ContainerCNIPause,
+		apicontainer.ContainerNamespacePause,
+		apicontainer.ContainerServiceConnectRelay,
+		apicontainer.ContainerManagedDaemon:
+		// pause images and AppNet relay image are managed at startup
+		return dockerapi.DockerContainerMetadata{}
+	}
+	// AppNet Agent container image is also managed at start up
+	// (it uses the same image as AppNet Relay container)
+	if task.IsServiceConnectEnabled() && container == task.GetServiceConnectContainer() {
+		return dockerapi.DockerContainerMetadata{}
+	}
+
+	var imageManifestDigest digest.Digest
+	digestFromPayload := referenceutil.GetDigestFromImageRef(container.Image)
+	if digestFromPayload != "" {
+		// Digest available in task payload
+		imageManifestDigest = digestFromPayload
+		logger.Info("Found image manifest digest in task payload", logger.Fields{
+			field.TaskARN:       task.Arn,
+			field.ContainerName: container.Name,
+			field.Image:         container.Image,
+		})
+	} else if !engine.imagePullRequired(engine.cfg.ImagePullBehavior, container, task.GetID()) {
+		// Digest should be resolved by inspecting local image
+		imgInspect, err := engine.client.InspectImage(container.Image)
+		if err != nil {
+			logger.Error("Failed to inspect image to find repo digest", logger.Fields{
+				field.TaskARN:       task.Arn,
+				field.ContainerName: container.Name,
+				field.Image:         container.Image,
+			})
+			return dockerapi.DockerContainerMetadata{
+				Error: dockerapi.CannotPullImageManifestError{FromError: err},
+			}
+		}
+		if len(imgInspect.RepoDigests) == 0 {
+			// Image was not pulled from a registry, so the user must have cached it on the
+			// host directly. Skip digest resolution for this case as there is no digest.
+			logger.Warn("No repo digest found", logger.Fields{
```

Comment:
If this is expected for some situations, maybe it should just be a "Info" message?

Reply:
Yeah.. I thought the usecase of a customer manually putting in an image on the host without it going through a pull is fairly niche so I chose warn. But since the usecase is legit I will change it to info.

Reply:
Updated.